File: cpp/560.subarray-sum-equals-k.cpp
/*
 * @lc app=leetcode id=560 lang=cpp
 *
 * [560] Subarray Sum Equals K
 */
#include <vector>
#include <unordered_map>
using namespace std;
// @lc code=start
class Solution
{
public:
    int subarraySum(vector<int> &nums, int k)
    {
        unordered_map<int, int> seen;
        int sum_val = 0;
        seen[sum_val] = 1;
        int res = 0;
        for (auto n : nums)
        {
            sum_val += n;
            if (seen.count(sum_val - k))
            {
                res += seen[sum_val - k];
            }
            ++seen[sum_val];
        }
        return res;
    }
};
// @lc code=end

File: java/560.subarray-sum-equals-k.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=560 lang=java
 *
 * [560] Subarray Sum Equals K
 */

// @lc code=start
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> seen = new HashMap<>();
        int sum_val = 0;
        seen.put(0, 1);
        int res = 0;
        for (int n : nums) {
            sum_val += n;
            if (seen.containsKey(sum_val - k)) {
                res += seen.get(sum_val - k);
            }
            seen.put(sum_val, seen.getOrDefault(sum_val, 0) + 1);
        }
        return res;
    }
}
// @lc code=end
